### DISCUSSION TITLE
LDAP support for TNS entries searching in addition to tnsnames.ora file

### DISCUSSION THREAD

**Discussion by danidorbek** [oracle/python-oracledb#71] [Ideas]
Hi,

it would be awesome if TNS alias could be queried from LDAP in THIN mode similarly how it works in Java Thin driver.
For example:
`
conn = oracledb.connect(user=usr, password=pwd, dsn='ldap://my_ldap.server/MY_TNS_ALIAS,cn=OracleContext,dc=sample,dc=com')
`

That would help to support large scale setups that use Oracle name resolution based on LDAP.

**Comment by cjbj** [comment]
Thanks for letting us know; it helps us prioritize features

**Reply by danidorbek** [reply]
it could be done using something like:

`

	def solve_ldap(ldap_dsn: str):
	    """ get TNS connect string from ldap.
	    Dependencies: re, ldap3

	    Input(str) : Input string in format "ldap://ldapserver/dbname,cn=OracleContext,dc=dom,dc=com"
	    returns(str): tns descriptor 

	    """
	    # get ldap parameters from dsn
	    import re
	    pattern = r"^ldap:\/\/(.+)\/(.+)\,(cn=OracleContext.*)$"
	    x = re.match(pattern, ldap_dsn)
	    if not x:
	        return None
	    else:
	        ldap_server, db, ora_context = x.groups()

	    from ldap3 import Server, Connection, ALL    
	    server = Server(ldap_server)
	    conn = Connection(server)
	    conn.bind()

	    conn.search(ora_context, f"(cn={db})", attributes=['orclNetDescString'])
	    tns = conn.entries[0].orclNetDescString.value
	    return tns
`

it does not take into account that there might be many LDAP servers (like java supports) but it works for start with simple case returning string.

**Reply by cjbj** [reply]
Thanks for sharing.

**Reply by mkmoisen** [reply]
Hi @danidorbek , this appears to work for me, thanks!

Would you please elaborate on the potential downsides to this approach? For example you mentioned there might be many LDAP servers. At least at my company there is just a single LDAP server hostname that we connect to, which I assume is load balanced but I don't know. 

Does your company have multiple LDAP servers and in Java do you pass them all into the JDBC connect string such that the client is responsible for trying each LDAP server until it finds a successful one?

**Reply by thepegasos** [reply]
This worked like a charm. Absolute genius!

**Reply by anthony-tuininga** [reply]
See the latest comment on #111 for support for registering a function for handling LDAP.

**Comment by cjbj** [comment]
Close this discussion enhancement: we'll track it in https://github.com/oracle/python-oracledb/issues/111
